File: Player/src/main/java/ru/testsimpleapps/coloraudioplayer/managers/player/data/StrictQueue.java
package ru.testsimpleapps.coloraudioplayer.managers.player.data;

import android.support.annotation.Nullable;

import java.util.LinkedList;

public final class StrictQueue<T> {

    private final static int DEFAULT_SIZE = 64;
    private final LinkedList<T> mQueue;
    private int mMaxSize = DEFAULT_SIZE;

    public StrictQueue() {
        mQueue = new LinkedList<>();
    }

    public StrictQueue(int maxSize) {
        mMaxSize = maxSize;
        mQueue = new LinkedList<>();
    }

    public void push(@Nullable T value) {
        if (mQueue.size() >= mMaxSize) {
            mQueue.removeFirst();
        }

        mQueue.addLast(value);
    }

    @Nullable
    public T pop() {
        return mQueue.pollLast();
    }

}

File: player/src/main/java/ru/testsimpleapps/coloraudioplayer/managers/player/data/StrictQueue.kt
package ru.testsimpleapps.coloraudioplayer.managers.player.data

import java.util.LinkedList

class StrictQueue<T> {
    private val mQueue: LinkedList<T>
    private var mMaxSize = DEFAULT_SIZE

    constructor() {
        mQueue = LinkedList()
    }

    constructor(maxSize: Int) {
        mMaxSize = maxSize
        mQueue = LinkedList()
    }

    fun push(value: T?) {
        if (mQueue.size >= mMaxSize) {
            mQueue.removeFirst()
        }

        mQueue.addLast(value)
    }

    fun pop(): T? {
        return mQueue.pollLast()
    }

    companion object {

        private val DEFAULT_SIZE = 64
    }

}
